Build failure: pkgsCross.mingwW64.openssl

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

`nix build github:NixOS/nixpkgs/nixos-unstable#pkgsCross.mingwW64.openssl`

### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
error: Cannot build '/nix/store/jsqmcbvfkhpipjq6r4fk7yw7m7kdlxwd-openssl-x86_64-w64-mingw32-3.6.0.drv'.
       Reason: builder failed with exit code 2.
       Output paths:
         /nix/store/0klijksm3ngj9bm1mhkidbfg4q2gpw03-openssl-x86_64-w64-mingw32-3.6.0-dev
         /nix/store/a5x6d5bas8452dms73dyrbf9f89w30r6-openssl-x86_64-w64-mingw32-3.6.0-bin
         /nix/store/gsscd7bp0igyn45lrlix4dqgyj7c7apr-openssl-x86_64-w64-mingw32-3.6.0
         /nix/store/m0r02h3b62384ha794q713nh1ibh68a7-openssl-x86_64-w64-mingw32-3.6.0-man
         /nix/store/sdgdgjaj41szv3x008qxvdllhw7y6r0l-openssl-x86_64-w64-mingw32-3.6.0-doc
       Last 25 log lines:
       > /nix/store/yyihdfl2r9akqx021bz1k13kflbc5cry-x86_64-w64-mingw32-binutils-2.44/bin/x86_64-w64-mingw32-ld: test/bioprinttest-bin-bioprinttest.obj:bioprinttest.c:(.text+0xc66): undefined reference to `__imp__get_printf_count_output'
       > rm -f test/conf_include_test.exe
       > rm -f test/confdump.exe
       > ${LDCMD:-x86_64-w64-mingw32-gcc} -m64 -Wall -O3 -L.   \
       >      -o test/conf_include_test.exe \
       >         test/conf_include_test-bin-conf_include_test.obj \
       >      test/libtestutil.a -lcrypto -lws2_32 -lgdi32 -lcrypt32
       > ${LDCMD:-x86_64-w64-mingw32-gcc} -m64 -Wall -O3 -L.   \
       >  -o test/cmsapitest.exe \
       >        test/cmsapitest-bin-cmsapitest.obj \
       >    test/libtestutil.a -lcrypto -lws2_32 -lgdi32 -lcrypt32
       > ${LDCMD:-x86_64-w64-mingw32-gcc} -m64 -Wall -O3 -L.   \
       >  -o test/confdump.exe \
       >  test/confdump-bin-confdump.obj \
       >        -lcrypto -lws2_32 -lgdi32 -lcrypt32
       > collect2: error: ld returned 1 exit status
       > rm -f test/constant_time_test.exe
       > ${LDCMD:-x86_64-w64-mingw32-gcc} -m64 -Wall -O3 -L.   \
       >        -o test/constant_time_test.exe \
       >        test/constant_time_test-bin-constant_time_test.obj \
       >    test/libtestutil.a -lcrypto -lws2_32 -lgdi32 -lcrypt32
       > make[1]: *** [Makefile:28271: test/bioprinttest.exe] Error 1
       > make[1]: *** Waiting for unfinished jobs....
       > make[1]: Leaving directory '/build/openssl-3.6.0'
       > make: *** [Makefile:4071: build_sw] Error 2
       For full logs, run:
         nix log /nix/store/jsqmcbvfkhpipjq6r4fk7yw7m7kdlxwd-openssl-x86_64-w64-mingw32-3.6.0.drv
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.53, NixOS, 25.11 (Xantusia), 25.11.20251019.5e2a59a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - nixpkgs: `/nix/store/7agp54mgffm9m1wc1kgmkm37pvy18qhf-source`

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/openssl/3.5/fix-mingw-linking.patch
+++ b/pkgs/development/libraries/openssl/3.5/fix-mingw-linking.patch
@@ -1,0 +1,33 @@
+From af3a3f8205968f9e652efa7adf2a359f4eb9d9cc Mon Sep 17 00:00:00 2001
+From: Alexandr Nedvedicky <sashan@openssl.org>
+Date: Mon, 6 Oct 2025 09:33:09 +0200
+Subject: [PATCH] OPENSSL_SYS_WINDOWS is also enabled for mingw build
+
+the test_n() in bioprinttest.c must differentiate between
+MSVC runtime libc and ming runtime libc. The function
+_set_printf_count_output() must be called when openssl
+is linked with MSVC libc only.
+
+Fixes #28679
+
+Reviewed-by: Bernd Edlinger <bernd.edlinger@hotmail.de>
+Reviewed-by: Tom Cosgrove <tom.cosgrove@arm.com>
+Reviewed-by: Neil Horman <nhorman@openssl.org>
+(Merged from https://github.com/openssl/openssl/pull/28759)
+---
+ test/bioprinttest.c | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/test/bioprinttest.c b/test/bioprinttest.c
+index bd99b9820ecc0..28730631bc201 100644
+--- a/test/bioprinttest.c
++++ b/test/bioprinttest.c
+@@ -541,7 +541,7 @@ static int test_n(int i)
+         ptrdiff_t t;
+     } n = { 0 };
+ 
+-#if defined(OPENSSL_SYS_WINDOWS)
++#if defined(_set_printf_count_output)
+     /*
+      * MS CRT is special and throws an exception when %n is used even
+      * in non-*_s versions of printf routines, and there is a special function

--- a/pkgs/development/libraries/openssl/default.nix
+++ b/pkgs/development/libraries/openssl/default.nix
@@ -422,6 +422,9 @@ in
         else
           ./3.5/use-etc-ssl-certs.patch
       )
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isMinGW [
+      ./3.5/fix-mingw-linking.patch
     ];
 
     withDocs = true;

